# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/CanvasChild.cpp
# Commit: 3f82f965831a
# Full Hash: 3f82f965831ac9966c8592e62b16c113492667b9
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2023-12-13 09:27:26
# Description:
#   Bug 1869661 - Add more null checks for CanvasChild::mRecorder. r=gfx-reviewers,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D196233
# ==============================================================================

diff -r de73ddb24c16 -r 3f82f965831a gfx/layers/ipc/CanvasChild.cpp
--- a/gfx/layers/ipc/CanvasChild.cpp	Wed Dec 13 03:41:19 2023 +0000
+++ b/gfx/layers/ipc/CanvasChild.cpp	Wed Dec 13 03:45:15 2023 +0000
@@ -259,7 +259,7 @@
     RecordEvent(RecordedCanvasEndTransaction());
     mIsInTransaction = false;
     mDormant = false;
-  } else {
+  } else if (mRecorder) {
     // Schedule to drop free buffers if we have no non-empty transactions.
     if (!mDormant) {
       mDormant = true;
@@ -275,12 +275,16 @@
 
 void CanvasChild::DropFreeBuffersWhenDormant() {
   // Drop any free buffers if we have not had any non-empty transactions.
-  if (mDormant) {
+  if (mDormant && mRecorder) {
     mRecorder->DropFreeBuffers();
   }
 }
 
-void CanvasChild::ClearCachedResources() { mRecorder->DropFreeBuffers(); }
+void CanvasChild::ClearCachedResources() {
+  if (mRecorder) {
+    mRecorder->DropFreeBuffers();
+  }
+}
 
 bool CanvasChild::ShouldBeCleanedUp() const {
   // Always return true if we've been deactivated.
@@ -289,11 +293,15 @@
   }
 
   // We can only be cleaned up if nothing else references our recorder.
-  return mRecorder->hasOneRef();
+  return !mRecorder || mRecorder->hasOneRef();
 }
 
 already_AddRefed<gfx::DrawTarget> CanvasChild::CreateDrawTarget(
     int64_t aTextureId, gfx::IntSize aSize, gfx::SurfaceFormat aFormat) {
+  if (!mRecorder) {
+    return nullptr;
+  }
+
   RefPtr<gfx::DrawTarget> dummyDt = gfx::Factory::CreateDrawTarget(
       gfx::BackendType::SKIA, gfx::IntSize(1, 1), aFormat);
   RefPtr<gfx::DrawTarget> dt = MakeAndAddRef<gfx::DrawTargetRecording>(
